Dispatching custom events (#201)
javascript-tutorial/it.javascript.info: 2-ui/2-events/05-dispatch-events/article.md

# Dispatching di eventi personalizzati

Oltre ad assegnare gestori, possiamo anche generare eventi attraverso JavaScript.

> Gli eventi personalizzati (custom events) possono essere usati per creare "componenti grafici". Ad esempio un elemento radice del nostro menù basato su JavaScript, può innescare degli eventi che indicano cosa avviene nel menù: `open` (se il menù è aperto), `select` (se c'è un elemento selezionato) e così via. Del codice in altre aree, potrebbe mettersi in ascolto per questi eventi e tenere traccia di cosa sta avvenendo nel menù.

Reviewer: Qui la traduzione sarebbe anche okay, ma non mi convince per nulla la frase, nemmeno il lingua originale a voler essere onesti... se si sta parlando di eventi, esempio "open" io lo intuisco come evento di qualcosa che è successo e non come "stato di apertura del menu".... infatti un evento come "click" indica che l'elemento ha subito un click, non sta dicendo se quell'elemento è già stato cliccato... Quindi ti propongo un suggerimento, ma è assolutamente opzionale.
```suggestion
Gli eventi personalizzati (custom events) possono essere usati per creare "componenti grafici". Ad esempio un elemento radice del nostro menù basato su JavaScript, potrebbe innescare eventi che indicano al menù come comportarsi: `open` (aprire il menu), `select` (selezionare un elemento) e così via. Del codice in altre aree, potrebbe mettersi in ascolto per questi eventi e tenere traccia di cosa sta avvenendo nel menù.
```

Author: /done

> Possiamo generare non solo eventi del tutto nuovi, creati apposta per i nostri scopi, ma anche quelli predefiniti del linguaggio, come `click`, `mousedown` etc. Questo può essere utile per dei testi automatici.

Reviewer: ```suggestion
Possiamo generare non solo eventi del tutto nuovi, creati apposta per i nostri scopi, ma anche quelli predefiniti del linguaggio, come `click`, `mousedown` etc. Questo può essere utile per i test automatici.
```

Author: /done

## Costruttore dell'evento 

Le classi degli eventi predefiniti formano una gerarchia, similmente alle classi degli elementi del DOM. La classe base è [Event](http://www.w3.org/TR/dom/#event).

Gli oggetti `Event` vengono creati così:

```js
let event = new Event(type[, options]);
```

Argomenti:

> - *type* -- tipo dell'evento, una stringa come `"click"` oppure una personale come `"mio-evento"`.

Reviewer: anche qui "personale" non mi piace troppo, nemmeno in inglese

```suggestion
- *type* -- tipo dell'evento, una stringa come `"click"` oppure una personalizzata come `"mio-evento"`.
```

Author: /done

- *options* -- l'oggetto con due proprietà opzionali:
  - `bubbles: true/false` -- se `true`, l'evento sarà soggetto a bubbling.
  - `cancelable: true/false` -- se `true`, "l'azione predefinita" può essere impedita. Successivamente vedremo cosa questo significhi per gli eventi personalizzati.

  Di default sono entrambi impostati a false: `{bubbles: false, cancelable: false}`.

## dispatchEvent

Dopo la creazione di un evento, dovremmo "eseguirlo" su un elemento, attraverso la chiamata `elem.dispatchEvent(event)`.

> A questo punto, il gestore reagisce su di esso come se fosse un evento del browser normale. Se l'evento è stato creato con l'opzione `bubbles` attivata, allora procederà con il bubbling.

Author: /done

```

```smart header="event.isTrusted"

> Per poter distinguere un evento generato da un utente "reale", da una generato da script esiste un modo.

Reviewer: ```suggestion
Per poter distinguere un evento generato da un utente "reale", da uno generato da script esiste un modo.
```

Author: /done

La proprietà `event.isTrusted` sarà `true` per eventi che provengono da azioni reali dell'utente e `false` per quelli generati da script.
```

## Esempio di bubbling

Possiamo creare un evento soggetto a bubbling con il nome `"hello"` e catturarlo attraverso `document`.

Tutto quello che dobbiamo fare è impostare `bubbles` a `true`:

```html run no-beautify
<h1 id="elem">Ciao dallo script!</h1>

<script>
  // catturato in document...
  document.addEventListener("hello", function(event) { // (1)
    alert("Ciao da " + event.target.tagName); // Ciao da H1
  });

  // ...generazione da elem!
  let event = new Event("hello", {bubbles: true}); // (2)
  elem.dispatchEvent(event);

  // il gestore su document si attiverà e visualizzerà il messaggio.

</script>
```

Note:

> 1. Per i nostri eventi personalizzati dobbiamo usare `addEventListener`, in quanto `on<event>` esiste solo per quelli predefiniti: `document.onhello` non funziona.

Reviewer: ```suggestion
1. Per i nostri eventi personalizzati dobbiamo usare `addEventListener`, in quanto `on<event>` esiste solo per quelli predefiniti: quindi `document.onhello` non funzionerà.
```

Author: /done

});

*!*
alert(event.clientX); // undefined, la proprietà sconosciuta viene ignorata!
*/!*
```

> Tecnicamente, possiamo aggirare il problema assegnandogli direttamente `event.clientX=100` dopo la creazione. Ma è una questione di comodità oltre che di aderenza alle regole. Gli eventi generati dal browser garantiscono sempre di avere sempre il tipo corretto.

Reviewer: ```suggestion
Tecnicamente, possiamo aggirare il problema assegnandogli direttamente `event.clientX=100` dopo la creazione. Ma è una questione di comodità oltre che di aderenza alle regole. Gli eventi generati dal browser garantiscono sempre di avere il tipo corretto.
```

Author: /done

> La lista completa delle proprietà, per i vari venti della UI sono descritti nelle specifiche, ad esempio, [MouseEvent](https://www.w3.org/TR/uievents/#mouseevent).

Reviewer: ```suggestion
La lista completa delle proprietà, per i vari eventi della UI sono descritti nelle specifiche, ad esempio, [MouseEvent](https://www.w3.org/TR/uievents/#mouseevent).
```

Author: /done

## Eventi personalizzati (custom events)

Per i nostri nuovi eventi come `"hello"` dovremmo usare `new CustomEvent`. Tecnicamente [CustomEvent](https://dom.spec.whatwg.org/#customevent) è equivalente a `Event`, senza eccezioni.

Nel secondo argomento (object) possiamo aggiungere la proprietà aggiuntiva `detail` per qualunque informazione personalizzata che vogliamo passare insieme all'evento.

Esempio:

```html run refresh
<h1 id="elem">Ciao a Giovanni!</h1>

<script>
  // insieme all'evento, al gestore, arrivano anche dei dettagli aggiuntivi
  elem.addEventListener("hello", function(event) {
    alert(*!*event.detail.name*/!*);
  });

  elem.dispatchEvent(new CustomEvent("hello", {
*!*
    detail: { name: "Giovanni" }
*/!*
  }));
</script>
```

La proprietà `detail` può contenere qualunque tipo di dato. Tecnicamente si potrebbe farne a meno, dato che possiamo assegnare qualunque proprietà dentro un normalissimo oggetto `new Event`, dopo la sua creazione. Ma `CustomEvent` fornisce il campo speciale `detail` adatto allo scopo, con cui ci si assicura di evitare conflitti con le altre proprietà dell'evento.

> Oltretutto, la classe evento descrive "che tipo di evento" sia, e se si tratta di un evento personalizzato, allora possiamo usare `CustomEvent` anche solo per palesare di cosa si tratti.

Author: /done

</script>
```

Nota bene: l'evento deve avere il flag `cancelable: true`, altrimenti la chiamata `event.preventDefault()` verrà ignorata.

> ## Gli eventi negli eventi sono sincroni

Reviewer: Anche qui, proposta totalmente opzionale
```suggestion
## Gli eventi annidati sono sincroni
```

Author: /done

    alert(2);
  };

  // viene innescato tra 1 e 2
  document.addEventListener('menu-open', () => alert('annidato'));
</script>
```

L'ordine di output è: 1 -> annidato -> 2.

Nota bene che l'evento annidato `menu-open` viene catturato nel `document`. La propagazione e la gestione dell'evento annidato vengono eseguiti e completati prima che l'elaborazione torni al codice esterno (`onclick`).

Questo non vale solo per `dispatchEvent`, ma esistono altri casi. Se un gestore di evento chiama metodi che innescano altri eventi -- anche questi vengono elaborati in maniera sincrona, in modo annidato.

Ora mettiamo il caso che questa cosa non ci stia bene, e che invece volessimo che `onclick` venisse elaborato per primo, indipendentemente da `menu-open` o prima di qualunque altro evento interno.

> Allora potremmo sia inserire `dispatchEvent` (o un altra chiamata che generi un evento) alla fine di  `onclick` o, forse anche meglio, avvolgerlo in un `setTimeout` a ritardo zero:

Reviewer 2: ```suggestion
Allora potremmo sia inserire `dispatchEvent` (o un altra chiamata che generi un evento) alla fine di  `onclick` oppure, forse anche meglio, incapsularlo in un `setTimeout` a ritardo zero:
```

Author: /done

</script>
```

> Adesso `dispatchEvent` viene eseguito in maniera asincrona, dopo che l'esecuzione del codice corrente è terminata, incluso `menu.onclick`, ed in questo modo i gestori sono totalmente separati.

Reviewer 2: ```suggestion
Adesso `dispatchEvent` viene eseguito in maniera asincrona dopo che l'esecuzione del codice corrente è terminata, incluso `menu.onclick`, ed in questo modo i gestori sono totalmente separati.
```

Author: /done

L'ordine di output diventa: 1 -> 2 -> annidato.

## Riepilogo

Per generare un evento dal codice, dobbiamo prima di tutto, creare un oggetto evento.

Il costruttore generico `Event(name, options)` accetta un nome di evento arbitrario e un oggetto `options`, con due proprietà:
- `bubbles: true` se l'evento deve fare bubbling.
- `cancelable: true` se `event.preventDefault()` deve poter funzionare.

> Altri costruttori di eventi nativi come `MouseEvent`, `KeyboardEvent` e così via, accettano proprietà specifiche per quel tipo di evento. Ad esempio, `clientX` per glie eventi del mouse.

Reviewer 2: ```suggestion
Altri costruttori di eventi nativi come `MouseEvent`, `KeyboardEvent` e così via, accettano proprietà specifiche per quel tipo di evento. Ad esempio, `clientX` per gli eventi del mouse.
```

Author: /done

Per eventi personalizzati dovremmo usare il costruttore `CustomEvent`, il quale ha una opzione aggiuntiva chiamata `detail`, per potervi assegnare i dati specifici del tipo di evento, ai quali i gestori potranno accedere tramite `event.detail`.

Nonostante tecnicamente esista la possibilità di generare eventi del browser come `click` o `keydown`, dovremmo usarli con molta attenzione.

> Non dovremmo generare eventi del browser perché una pratica poco elegante e sporca nell'esecuzione dei gestori. La maggior parte delle volte, si tratta di cattiva architettura.

Reviewer: ```suggestion
Non dovremmo generare eventi del browser perché è una pratica poco elegante e sporca nell'esecuzione dei gestori. La maggior parte delle volte, si tratta di cattiva architettura.
```

Author: /done

> Gi eventi nativi possono essere generati quando:

Reviewer 2: ```suggestion
Gli eventi nativi possono essere generati:
```

Author: /done
